### PR TITLE
[FLINK-13281] Fix the verify logic of AdaptedRestartPipelinedRegionStrategyNGAbortPendingCheckpointsTest

### PR DESCRIPTION
**Pull request by Myasuka**
## What is the purpose of the change

[FLINK-13060](https://issues.apache.org/jira/browse/FLINK-13060) ensure FailoverStrategies should respect restart constraints. However, this change did not cover the `AdaptedRestartPipelinedRegionStrategyNGAbortPendingCheckpointsTest` which should also set as `FixedDelayRestartStrategy`. Currently, as the pending checkpoint would still be discarded due to `failUnacknowledgedPendingCheckpoints` within `CheckpointCoordinator`, `AdaptedRestartPipelinedRegionStrategyNGAbortPendingCheckpointsTest` could pass the test.

This is actually not what we want the test did, we should actually verify that when region failover, pending checkpoints would be discarded.

## Brief change log

* Fix the verify logic of `AdaptedRestartPipelinedRegionStrategyNGAbortPendingCheckpointsTest`

## Verifying this change

This change modify previous test `AdaptedRestartPipelinedRegionStrategyNGAbortPendingCheckpointsTest`

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): **no**
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: **no**
  - The serializers: **no**
  - The runtime per-record code paths (performance sensitive): **no**
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: **no**
  - The S3 file system connector: **no**

## Documentation

  - Does this pull request introduce a new feature? **no**
  - If yes, how is the feature documented? **not applicable**

### CODE REVIEW

**Comment by Myasuka** [comment]
@flinkbot attention @zentol

**Comment by tillrohrmann** [comment]
Why do you think that we no longer correctly test that pending checkpoints are being aborted with FLINK-13060 @Myasuka?

**Comment by Myasuka** [comment]
@tillrohrmann I would try to interpret this more clearly below:
After [FLINK-13060](https://issues.apache.org/jira/browse/FLINK-13060), `AdaptedRestartPipelinedRegionStrategyNGAbortPendingCheckpointsTest` still use previous `InfiniteDelayRestartStrategy` which means `AdaptedRestartPipelinedRegionStrategyNG#resetTasks()` would never be called when some task failed. The root cause why previous `AdaptedRestartPipelinedRegionStrategyNGAbortPendingCheckpointsTest` could still pass is because of the pending checkpoints are aborted by `failUnacknowledgedPendingCheckpointsFor()` method in `CheckpointCoordinator` (which is actually introduced vai [FLINK-7844](https://issues.apache.org/jira/browse/FLINK-7844) ). 

In a nutshell, we expect pending checkpoints are aborted by `AdaptedRestartPipelinedRegionStrategyNG#resetTasks()` not `CheckpointCoordinator#failUnacknowledgedPendingCheckpointsFor()`.

**Comment by zentol** [comment]
Why are you adding a second vertex?

**Comment by Myasuka** [comment]
@zentol If only one vertex, after that vertex sent acknowledge message to checkpoint coordinator and that checkpoint would be treated as completed checkpoint. Actually I just want that checkpoint still stay as pending checkpoint.

File: flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/AdaptedRestartPipelinedRegionStrategyNGAbortPendingCheckpointsTest.java [hunk 1]
```diff
@@ -76,49 +80,64 @@ public void abortPendingCheckpointsWhenRestartingTasks() throws Exception {
...
 		final int pendingCheckpointsBeforeFailure = checkpointCoordinator.getNumberOfPendingCheckpoints();
+		long checkpointId = checkpointCoordinator.getPendingCheckpoints().keySet().iterator().next();
 
-		failVertex(onlyExecutionVertex);
+		AcknowledgeCheckpoint acknowledgeCheckpoint = new AcknowledgeCheckpoint(
+			firstExecutionVertex.getJobId(),
+			firstExecutionVertex.getCurrentExecutionAttempt().getAttemptId(),
+			checkpointId,
+			mock(CheckpointMetrics.class),
+			mock(TaskStateSnapshot.class));
```
Comment (zentol) [inline flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/AdaptedRestartPipelinedRegionStrategyNGAbortPendingCheckpointsTest.java]: mocking isn't necessary, we can easily create instances for both of these.

File: flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/AdaptedRestartPipelinedRegionStrategyNGAbortPendingCheckpointsTest.java [hunk 2]
```diff
@@ -76,49 +80,64 @@ public void abortPendingCheckpointsWhenRestartingTasks() throws Exception {
...
+
+		// not let the first checkpoint to be discarded when firstExecutionVertex has been acknowledged.
+		checkpointCoordinator.receiveAcknowledgeMessage(acknowledgeCheckpoint, "Unknown location");
```
Comment (zentol) [inline flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/AdaptedRestartPipelinedRegionStrategyNGAbortPendingCheckpointsTest.java]: add assertion / state check that after this we still have a pending checkpoint

File: flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/AdaptedRestartPipelinedRegionStrategyNGAbortPendingCheckpointsTest.java [hunk 3]
```diff
@@ -76,49 +80,64 @@ public void abortPendingCheckpointsWhenRestartingTasks() throws Exception {
...
 
-		setTaskRunning(executionGraph, onlyExecutionVertex);
+		setTaskRunning(executionGraph, firstExecutionVertex, vertexIterator.next());
 
 		final CheckpointCoordinator checkpointCoordinator = executionGraph.getCheckpointCoordinator();
 		checkState(checkpointCoordinator != null);
 
 		checkpointCoordinator.triggerCheckpoint(System.currentTimeMillis(),  false);
 		final int pendingCheckpointsBeforeFailure = checkpointCoordinator.getNumberOfPendingCheckpoints();
+		long checkpointId = checkpointCoordinator.getPendingCheckpoints().keySet().iterator().next();
```
Comment (zentol) [inline flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/AdaptedRestartPipelinedRegionStrategyNGAbortPendingCheckpointsTest.java]: add `checkState(pendingCheckpointsBeforeFailure == 1)`, remove assertion at the end

File: flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/AdaptedRestartPipelinedRegionStrategyNGAbortPendingCheckpointsTest.java [hunk 4]
```diff
@@ -76,49 +80,64 @@ public void abortPendingCheckpointsWhenRestartingTasks() throws Exception {
...
 
-		failVertex(onlyExecutionVertex);
+		AcknowledgeCheckpoint acknowledgeCheckpoint = new AcknowledgeCheckpoint(
+			firstExecutionVertex.getJobId(),
```
Comment (zentol) [inline flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/AdaptedRestartPipelinedRegionStrategyNGAbortPendingCheckpointsTest.java]: retrieve the job ID from the execution graph instead

File: flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/AdaptedRestartPipelinedRegionStrategyNGAbortPendingCheckpointsTest.java [hunk 5]
```diff
@@ -76,49 +80,64 @@ public void abortPendingCheckpointsWhenRestartingTasks() throws Exception {
...
+			firstExecutionVertex.getCurrentExecutionAttempt().getAttemptId(),
+			checkpointId,
+			mock(CheckpointMetrics.class),
+			mock(TaskStateSnapshot.class));
+
+		// not let the first checkpoint to be discarded when firstExecutionVertex has been acknowledged.
```
Comment (zentol) [inline flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/AdaptedRestartPipelinedRegionStrategyNGAbortPendingCheckpointsTest.java]:
```suggestion
		// let the first vertex acknowledge the checkpoint, and fail it afterwards
		// the failover strategy should then cancel all pending checkpoints on restart
```

File: flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/AdaptedRestartPipelinedRegionStrategyNGAbortPendingCheckpointsTest.java [hunk 6]
```diff
@@ -76,49 +80,64 @@ public void abortPendingCheckpointsWhenRestartingTasks() throws Exception {
...
+				new TaskExecutionState(executionGraph.getJobID(),
+					executionVertex.getCurrentExecutionAttempt().getAttemptId(),
+					ExecutionState.RUNNING));
+		}
 	}
 
 	private void failVertex(final ExecutionVertex onlyExecutionVertex) {
 		onlyExecutionVertex.getCurrentExecutionAttempt().fail(new Exception("Test Exception"));
 		manualMainThreadExecutor.triggerAll();
+		manualMainThreadExecutor.triggerScheduledTasks();
```
Comment (zentol) [inline flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/AdaptedRestartPipelinedRegionStrategyNGAbortPendingCheckpointsTest.java]: We could do this later after an assertion that we still have a pending checkpoint; this would give use more confidence that the actual restart by the failover strategy discarded the pending checkpoints

File: flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/AdaptedRestartPipelinedRegionStrategyNGAbortPendingCheckpointsTest.java [hunk 7]
```diff
@@ -76,49 +80,64 @@ public void abortPendingCheckpointsWhenRestartingTasks() throws Exception {
...
 		assertThat(pendingCheckpointsBeforeFailure, is(equalTo(1)));
 		assertNoPendingCheckpoints(checkpointCoordinator);
 	}
 
-	private void setTaskRunning(final ExecutionGraph executionGraph, final ExecutionVertex executionVertex) {
-		executionGraph.updateState(
-			new TaskExecutionState(executionGraph.getJobID(),
-				executionVertex.getCurrentExecutionAttempt().getAttemptId(),
-				ExecutionState.RUNNING));
+	private void setTaskRunning(final ExecutionGraph executionGraph, final ExecutionVertex... executionVertices) {
```
Comment (zentol) [inline flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/AdaptedRestartPipelinedRegionStrategyNGAbortPendingCheckpointsTest.java]: `setTasksRunning`

File: flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/AdaptedRestartPipelinedRegionStrategyNGAbortPendingCheckpointsTest.java (hunk 6 above)
Comment (zentol) [reply flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/AdaptedRestartPipelinedRegionStrategyNGAbortPendingCheckpointsTest.java]: Where did you resolve this? My goal was that we verified that a pending checkpoint still exists after `ExecutionAttempt#fail`, but before the actual restart by the failover strategy.

Comment (Myasuka) [reply flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/AdaptedRestartPipelinedRegionStrategyNGAbortPendingCheckpointsTest.java]: fixed in latest commit.
